Extend WKD support to include WKD URLs

wiktor-k opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Description
This change allows handling Web Key Directory links by OpenKeychain (they always have scheme `https` and path starting with `/.well-known/openpgpkey/hu`).

This change also fixes two issues with WKD: 1) handling e-mails with leading/trailing spaces 2) following redirects on WKD links, just like gnupg does (some hosts like `kernel.org` use it). It also adds tests for WKD URL construction utilities.

## Motivation and Context
This change completes support of WKD in #2270 and allows importing keys over WKD using links similarly to how Facebook or HKP links are handled.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Deployed debug build on a phone and tested links mentioned in #2270.

## Screenshots (if appropriate):
After clicking on a link.
![screenshot_1526978160](https://user-images.githubusercontent.com/1718963/40353183-23ed0e42-5db1-11e8-9b1e-ed57ee6c247a.png)

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ Bug fix (non-breaking change which fixes an issue)
- ✅ New feature (non-breaking change which adds functionality)

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/util/WebKeyDirectory.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class WebKeyDirectory {
```
*Util

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java:
```diff
@@ -208,6 +212,13 @@ protected void handleActions(@NonNull Intent intent) {
                 break;
             }
+            case ACTION_IMPORT_KEY_FROM_WEB_KEY_DIRECTORY: {
+                Preferences.CloudSearchPrefs cloudSearchPrefs =
```
would be nicer to use a factory method here

wiktor-k replied:
Ah, that's what you meant, I thought this was about `autovalue` only ;) Okay, let me fix that....

wiktor-k replied:
Okay fixed, there are two new methods used now: `createWebKeyDirectoryOnly` and `createKeyserverOnly`.

wiktor-k commented:
Okay, now it should be better. I wanted to delay AutoValue until further PR but if you're insisting I added it here. Also, I've added the preference pane, the commit looks small. I hope this entire PR did not grow too big ;-)

Thanks for your review!

wiktor-k commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/util/Preferences.java:
```diff
@@ -362,62 +361,32 @@ public void setKeySignaturesTableInitialized() {
-        public final HkpKeyserverAddress keyserver;
+    @AutoValue
+    public static abstract class CloudSearchPrefs implements Parcelable {
+        public abstract boolean searchKeyserver();
```
I didn't have better name for these accessor methods... `isSearchKeyserver`? `isSearchingKeyserver`? `shouldSearchKeyserver`? If you've got better idea tell me and I'll adjust the code.

Valodim replied:
Perhaps isKeyserverEnabled?

wiktor-k commented:
Okay @Valodim, this should be it. Phew...

Valodim commented on OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/ImportKeysActivity.java:
```diff
@@ -203,17 +207,23 @@ protected void handleActions(@NonNull Intent intent) {
 
                 Preferences.CloudSearchPrefs cloudSearchPrefs =
-                        new Preferences.CloudSearchPrefs(false, true, true, false, null);
+                        Preferences.CloudSearchPrefs.create(false, true, true, false, null);
```
:)

Valodim commented:
Nice work, thanks!
